app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/uploadsupportingdocs/UploadSupportingEvidenceController.scala

```diff
@@ -0,0 +1,1012 @@
+                      Some(doYouWantToUploadSupportingEvidenceAnswer),
+                      supportingEvidences :+ SupportingEvidence(reference, supportingEvidence.filename)
+                    )
+                }
+                newDraftReturn = fillingOutReturn.draftReturn match {
+                  case s @ DraftSingleDisposalReturn(
+                        _,
+                        _,
+                        _,
+                        _,
+                        _,
+                        _,
+                        _,
+                        _,
+                        _,
+                        _,
+                        _
+                      ) =>
+                    s.copy(uploadSupportingDocuments = Some(updatedAnswers))
+                  case m @ DraftMultipleDisposalsReturn(
+                        _,
+                        _,
+                        _,
+                        _,
+                        _,
+                        _,
+                        _
+                      ) =>
+                    m.copy(uploadSupportingDocuments = Some(updatedAnswers))
+                }
+                _ <- returnsService
+                      .storeDraftReturn(
+                        newDraftReturn,
+                        fillingOutReturn.subscribedDetails.cgtReference,
+                        fillingOutReturn.agentReferenceNumber
+                      )
+                _ <- EitherT(
+                      updateSession(sessionStore, request)(
+                        _.copy(journeyStatus = Some(fillingOutReturn.copy(draftReturn = newDraftReturn)))
+                      )
+                    )
+
+              } yield reference
+
+              result.fold(
+                error => {
+                  logger.warn(s"failed to upload file with error: $error")
+                  errorHandler.errorResult(None)
+                },
+                ref => Redirect(routes.UploadSupportingEvidenceController.uploadSupportingEvidenceVirusCheck(ref))
+              )
+            }
+          )
+          .getOrElse {
+            logger.warn("missing file key")
+            Future.successful(errorHandler.errorResult(None))
+          }
+      }
+    }
+
+  def uploadSupportingEvidenceVirusCheck(reference: String): Action[AnyContent] =
+    authenticatedActionWithSessionData.async { implicit request =>
+      withUploadSupportingEvidenceAnswers(request) { (draftReturnId, _, _, _, _) =>
+        val result = for {
+          fd <- upscanService
+                 .getUpscanFileDescriptor(DraftReturnId(draftReturnId.toString), UpscanInitiateReference(reference))
+        } yield fd
+
+        result.fold(
+          error => {
+            logger.warn(s"failed to get file descriptor information: $error")
+            errorHandler.errorResult(None)
+          }, {
+            case Some(upscanFileDescriptor) =>
+              if (upscanFileDescriptor.status === UPLOADED | upscanFileDescriptor.status === FAILED) {
```

Comment:
use two `||` to get the short-circuiting behaviour

Reply:
what about the `READY` state? The template seems to handle that status

Reply:
READY is handled by the else branch - there are only three states to consider.